ipn/ipnlocal/web.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package ipnlocal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"tailscale.com/client/tailscale"
+	"tailscale.com/client/web"
+	"tailscale.com/envknob"
+)
+
+// webServer holds state for the web interface for managing
+// this tailscale instance. The web interface is not used by
+// default, but initialized by calling LocalBackend.WebOrInit.
+type webServer struct {
+	ws         *web.Server  // or nil, initialized lazily
+	httpServer *http.Server // or nil, initialized lazily
+
+	// lc optionally specifies a LocalClient to use to connect
+	// to the localapi for this tailscaled instance.
+	// If nil, a default is used.
+	lc *tailscale.LocalClient
+
+	wg sync.WaitGroup
+}
+
+// SetWebLocalClient sets the b.web.lc function.
+// If lc is provided as nil, b.web.lc is cleared out.
+func (b *LocalBackend) SetWebLocalClient(lc *tailscale.LocalClient) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.web.lc = lc
+}
+
+// WebOrInit gets or initializes the web interface for
+// managing this tailscaled instance.
+func (b *LocalBackend) WebOrInit() (_ *web.Server, err error) {
+	if !envknob.Bool("TS_DEBUG_WEB_UI") {
+		return nil, errors.New("web ui flag unset")
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.web.ws != nil {
+		return b.web.ws, nil
+	}
+
+	b.logf("WebOrInit: initializing web ui")
+	if b.web.ws, err = web.NewServer(web.ServerOpts{
+		// TODO(sonia): allow passing back dev mode flag
+		LocalClient:    b.web.lc,
+		DoNoiseRequest: b.DoNoiseRequest,
+		Logf:           b.logf,
+	}); err != nil {
+		return nil, fmt.Errorf("web.NewServer: %w", err)
+	}
+
+	// Start up the server.
+	b.web.wg.Add(1)
+	go func() {
+		defer b.web.wg.Done()
+		addr := ":5252"
```

Comment:
I think for this client instance, we only want to listen on the Tailscale interface, don't we?

Reply:
In looking at the cleanup, I suspect that we will want to hook into [LocalBackend.TCPHandlerForDst](https://github.com/tailscale/tailscale/blob/8326fdd60f2596bb73e1ed7ef0b341c0f1f1b58f/ipn/ipnlocal/local.go#L3085) to intercept requests for the web client.  That will also be needed if and when we want to serve on quad100 anyway.

That will end up refactoring a good chunk of this code here, but I think that's alright.

Reply:
In looking at the cleanup, I suspect that we will want to hook into [LocalBackend.TCPHandlerForDst](https://github.com/tailscale/tailscale/blob/8326fdd60f2596bb73e1ed7ef0b341c0f1f1b58f/ipn/ipnlocal/local.go#L3085) to intercept requests for the web client (in the same way we do for ssh, serve, and peerapi).  That will also be needed if and when we want to serve on quad100 anyway.

That will end up refactoring a good chunk of this code here, but I think that's alright.